### PR TITLE
plugins.vlive: fix URL regex and plugin

### LINKED ISSUE #3314: vlive: plugin is broken
## Plugin Issue

- [x] This is a plugin issue and I have read the contribution guidelines.

### Description

VLive.tv plugin is broken due to revising the site. The way to get parameters required for getting stream urls is changed now.

### Reproduction steps / Explicit stream URLs to test

Example: streamlink https://www.vlive.tv/video/221427 best

### Log output

```
streamlink -l debug https://www.vlive.tv/video/221427
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.6.6
[cli][debug] Streamlink: 1.7.0
[cli][debug] Requests(2.24.0), Socks(1.7.1), Websocket(0.57.0)
[cli][info] Found matching plugin vlive for URL https://www.vlive.tv/video/221427
error: Could not find video status information!
```

### Additional comments, etc.

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

### PR DESCRIPTION
**Pull request by SeonjaeHyeon**
closes #3314 

The site was completely revised recently and url format was changed.
But old format still exists, so I added support for both formats.

### CODE REVIEW

**Comment by mkbloke** [comment]
I was just looking at this myself.  One thing I noticed is that the literal dot after www in the URL regex needs escaping, although that was present before your changes.

I know you've just followed what was previously done in the existing plugin, but it's possible when one of the main Streamlink devs reviews your code that you could be asked to change the `raise PluginError()`s to `log.error()`s and then return.  At least, that seems to be the way they prefer that I have seen.  I'm sure you'll find out soon enough from one of them though.

File: src/streamlink/plugins/vlive.py [hunk 1]
```diff
@@ -9,48 +9,60 @@
...
         # All "resolutions" have a variant playlist with only one entry, so just combine them
-        for i in stream_info['resolutions']:
-            res_streams = HLSStream.parse_variant_playlist(self.session, i['cdnUrl'])
+        for i in stream_info['result']['streamList']:
+            res_streams = HLSStream.parse_variant_playlist(self.session, i['serviceUrl'])
             if len(res_streams.values()) > 1:
                 log.warning('More than one stream in variant playlist, using first entry!')
 
-            streams[i['name']] = res_streams.popitem()[1]
+            streams[i['streamName']] = res_streams.popitem()[1]
```
Comment (bastimeyer) [inline src/streamlink/plugins/vlive.py:65]: Just yielding a tuple of the stream name and HLSStream instance is cleaner than adding them to a dict and returning it afterwards, but that's just a "cosmetic" code suggestion.

File: src/streamlink/plugins/vlive.py [hunk 2]
```diff
@@ -9,48 +9,60 @@
...
-            raise PluginError('Could not get video init page (HTTP Status {})'
-                              .format(vinit_req.status_code))
+        video_page = self.session.http.get(self.url, headers=dict(referer=self.url))
+        if video_page.status_code != 200:
+            raise PluginError('Could not get video page (HTTP Status {})'
+                              .format(video_page.status_code))
+
+        page_info_js = self._page_info.search(video_page.text)
+        if not page_info_js:
+            raise PluginError('Could not find page info')
```
Comment (bastimeyer) [inline src/streamlink/plugins/vlive.py:28]: Unavailable streams are not supposed to raise a PluginError. Just return None or an empty dict. PluginErrors are meant for actual errors. Same issue below...

**Comment by SeonjaeHyeon** [comment]
I appreciate your help. @mkbloke @bastimeyer 

I modified PR as you suggested to me.
But I'm not familiar with using schemas and yield, I'm not sure it is correct.

Plus, you can find Live streams here: https://www.vlive.tv/home/new?lang=en (it will be redirected to english page)
And you will see two types of video, Live and replay.
The live video has LIVE sign on thumbnail. Maybe you can find it easily.
There might be no live videos when you check the site due to time difference.

File: src/streamlink/plugins/vlive.py [hunk 3]
```diff
@@ -1,58 +1,85 @@
...
+from streamlink.utils import parse_json
 
 log = logging.getLogger(__name__)
 
 
 class Vlive(Plugin):
-    _url_re = re.compile(r"https?://(?:www.)vlive\.tv/video/(\d+)")
-    _video_status = re.compile(r'oVideoStatus = (.+)<', re.DOTALL)
-    _video_init_url = "https://www.vlive.tv/video/init/view"
+    _url_re = re.compile(r"https?://(?:www\.)vlive\.tv/(?P<format>video|post)/(?P<id>[0-9\-]+)")
```
Comment (back-to) [inline src/streamlink/plugins/vlive.py:13]:
```suggestion
    _url_re = re.compile(r"https://www\.vlive\.tv/(?P<format>video|post)/(?P<id>[0-9\-]+)")
```
if there is only https support you don't need to ignore the `s`
same for the `www` subdomain

File: src/streamlink/plugins/vlive.py [hunk 4]
```diff
@@ -1,58 +1,85 @@
...
+                                           schema=self._schema_video)
+        if video_json is None:
+            log.error('Could not parse video page')
+            return
+
+        if video_json['type'] == 'VOD':
+            log.error('VODs are not supported')
+            return
+
+        url_format, video_id = re.match(self._url_re, self.url).group('format', 'id')
```
Comment (back-to) [inline src/streamlink/plugins/vlive.py:60]:
```suggestion
        url_format, video_id = self._url_re.match(self.url).groups()
```

you don't need `re.`

**Review by bastimeyer — COMMENTED** [review]
Code-wise, this is looking good now. I still don't see any live streams on this site though, so I can't verify it.

**Review by bastimeyer — APPROVED** [review]

**Comment by gravyboat** [comment]
I'd like to see examples of actual live streams. I checked yesterday and today and didn't see anything live that I could actually watch. It just looks like the equivalent of a blogging platform. How often are there live streams? Where can these streams be viewed? Are the streams geo locked? Until these questions are answered this PR should not be merged, and if these questions cannot be answered and it's simply for watching VODs or live streams of pre-recorded content the plugin should instead be removed.

**Review by gravyboat — CHANGES_REQUESTED** [review]
See my above comment.

**Comment by SeonjaeHyeon** [comment]
> I'd like to see examples of actual live streams. I checked yesterday and today and didn't see anything live that I could actually watch. It just looks like the equivalent of a blogging platform.

As I said above, I think you have time difference problem. This site is a South Korean live streaming platform and many korean celebrities broadcast live videos. So, you should check the site considering korea time (GMT +9).

> How often are there live streams?

Because most live broadcasts are irregular, all I can say is, there were some live videos between evening and midnight.

And now, I can see two live videos.
https://imgur.com/xA5vsNO

> Where can these streams be viewed?

You can see the live videos the New page: https://www.vlive.tv/home/new?lang=en
or, on particular channel, check Star Board page.

> Are the streams geo locked?

I don't think they are geo-blocked. Fans from all over the world watch and chat when their celebrities broadcast.

File: src/streamlink/plugins/vlive.py [hunk 5]
```diff
@@ -1,58 +1,85 @@
...
+        validate.transform(_page_info.search),
+        validate.any(None, validate.all(
+            validate.get(1),
+            validate.transform(parse_json),
+            {"postDetail": {"post": {"officialVideo": {
+                "type": str,
+                "videoSeq": int,
+                validate.optional("status"): str,
+            }}}},
+            validate.get("postDetail"),
```
Comment (back-to) [inline src/streamlink/plugins/vlive.py:27]: you need to cache the errors

```
[cli][info] Found matching plugin vlive for URL https://www.vlive.tv/video/222194
error: Unable to validate response text: <re.Match object; span=(3665, 6426), match='window.__PRELOADED_STATE__={"common":{"userCountr> does not equal None or Unable to validate key 'postDetail': Key 'post' not found in 
{
    'error': {
        'errorCode': 'common_700',
        'message': 'Only members of this channel can continue. \nJoin 온앤오프 (ONF) channel!',
        'data': {
            'url': 'https://vlive.tv/post/0-19795109',
            'title': '안뇽 퓨즈🌙',
            'createdAt': 1604680573000,
            'availableActions': [],
            'commentCount': 0,
            'emotionCount': 0,
            'postId': '0-19795109',
            'isCommentEnabled': True,
            'officialVideo': {
                'videoSeq': 222194,
                'type': 'LIVE',
                'title': '안뇽 퓨즈🌙',
                'multinationalTitles': [{
                    'type': 'VIDEO_TITLE',
                    'seq': 222194,
                    'locale': 'ko_KR',
                    'label': '안뇽 퓨즈🌙',
                    'defaultYn': True
                }],
                'playCount': 0,
                'likeCount': 387749,
                'commentCount': 7470,
                'thumb': 'http://thumb.vlive.tv/live/222194/thumb',
                'exposeStatus': 'EXPOSED',
                'screenOrientation': 'HORIZONTAL',
                'willStartAt': 1604680572000,
                'onAirStartAt': 1604680573000,
                'willEndAt': 1604681498000,
                'createdAt': 1604680572000,
                'liveThumbYn': False,
                'upcomingYn': False,
                'productType': 'NONE',
                'preAdYn': True,
                'postAdYn': False,
                'mobileDAYn': True,
                'vrContentType': 'NONE',
                'schemaVersion': 'V202009',
                'liveType': 'GENERAL',
                'status': 'ENDED',
                'startAt': 1604680591000,
                'endAt': 1604681498000,
                'watchedCount': 9813,
                'publicYn': False,
                'userSeq': 14376571,
                'momentable': False
            },
            'contentType': 'VIDEO',
            'author': {
                'memberId': 'a6261a101b7192d90942867e6027dde3',
                'channelCode': 'E812E3',
                'joined': True,
                'nickname': 'Staff_온앤오프 (ONF)',
                'profileImageUrl': 'http://v.phinf.naver.net/20200808_114/1596822845075mGyrT_PNG/profile370_83164.png',
                'officialProfileType': 'STAFF'
            },
            'channel': {
                'channelCode': 'E812E3',
                'channelName': '온앤오프 (ONF)'
            },
            'board': {
                'boardId': 7020,
                'title': '[ONF FUSE 1ST] Star Board',
                'boardType': 'STAR',
                'payRequired': True,
                'readAllowedLabel': 'Who purchased ONF OFFICIAL FANSHIP FUSE 1ST'
            }
        }
    },
    'onLoading': False
}
```

**Review by back-to — CHANGES_REQUESTED** [review]

**Pull request merged**